app/models/contest_score.rb

```diff
@@ -19,10 +19,11 @@ def recalculate_and_save
         destroy # in case already in database - this occurs if submissions get deleted
       else
         self.attempts = attempts # attempts
-        submission = submissions.order("evaluation DESC, created_at ASC").first
-        self.attempt = submissions.where("created_at <= ?", submission.created_at).count # attempts number
+
+        weighting = contest.problem_set.problem_associations.find_by(problem_id: problem_id).weighting
+        unweighted_score, self.attempt, submission = ScoringMethods.score_problem_submissions(problem, submissions) # Does correct subtask or maximum scoring
+        self.score = (unweighted_score * weighting).to_i
```

Comment:
unweighted score _could_ be nil here, just because ScoringMethods.score_problem_submissions can be nil. It could be worth either defaulting to zero or maybe better just adding `if unweighted_score?` as a postfix if to this line?

Reply:
I agree it's pretty unlikely, but it feels worth doing just from a "making the system more resilient" point of view. I agree with your analysis on nil vs zero!